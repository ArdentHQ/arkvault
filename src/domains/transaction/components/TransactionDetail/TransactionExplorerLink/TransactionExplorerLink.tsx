import cn from "classnames";
import { DTO } from "@payvo/sdk-profiles";
import React, { useMemo } from "react";
import { useTranslation } from "react-i18next";

import { Clipboard } from "@/app/components/Clipboard";
import { Icon } from "@/app/components/Icon";
import { Link } from "@/app/components/Link";

import {
	TransactionDetail,
	TransactionDetailProperties,
} from "@/domains/transaction/components/TransactionDetail/TransactionDetail";
import { useBreakpoint } from "@/app/hooks";
import { TruncateMiddle } from "@/app/components/TruncateMiddle";

type TransactionExplorerLinkProperties = {
	isDisabled?: boolean;
	transaction: DTO.ExtendedConfirmedTransactionData | DTO.ExtendedSignedTransactionData;
} & TransactionDetailProperties;

export const TransactionExplorerLink = ({
	isDisabled,
	transaction,
	borderPosition = "top",
	...properties
}: TransactionExplorerLinkProperties) => {
	const { t } = useTranslation();
	const { isMdAndAbove, isXs } = useBreakpoint();

	const maxChars = useMemo(() => {
		if (isMdAndAbove) {
			return 56;
		}

		if (isXs) {
			return 18;
		}

		return 30;
	}, [isMdAndAbove]);

	return (
		<TransactionDetail label={t("TRANSACTION.ID")} borderPosition={borderPosition} {...properties}>
			<div className="flex w-full items-center space-x-3 overflow-hidden">
				<span className={cn("overflow-hidden", { "w-full": !isMdAndAbove })}>
					<Link
						className={cn("flex items-center", { "justify-end": !isMdAndAbove })}
						to={transaction.explorerLink()}
						isDisabled={isDisabled}
						isExternal
					>
						<TruncateMiddle text={transaction.id()} maxChars={maxChars} />
					</Link>
				</span>

				<span className="flex text-theme-primary-300 dark:text-theme-secondary-600">
					<Clipboard variant="icon" data={transaction.id()}>
						<Icon name="Copy" />
					</Clipboard>
				</span>
			</div>
		</TransactionDetail>
	);
};
