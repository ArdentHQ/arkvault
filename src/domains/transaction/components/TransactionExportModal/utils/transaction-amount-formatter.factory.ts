import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO, Helpers } from "@ardenthq/sdk-profiles";

const formatAmount = (amount: number, transaction: DTO.ExtendedConfirmedTransactionData) =>
	Helpers.Currency.format(amount, transaction.wallet().exchangeCurrency(), {
		withTicker: false,
	});

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isReceived()) {
		let totalReceived = BigNumber.make(transaction.amount());

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived.minus(recipient.amount);
			}
		}

		return formatAmount(totalReceived.toNumber(), transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return multiPaymentAmount(transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return BigNumber.make(multiPaymentAmount(transaction)).plus(transaction.fee());
	}

	return formatAmount(transaction.total(), transaction);
};

export const AmountFormatter = (transaction: DTO.ExtendedConfirmedTransactionData) => ({
	amount: () => transactionAmount(transaction),
	fee: () => formatAmount(transaction.fee(), transaction),
	total: () => transactionTotal(transaction),
});
