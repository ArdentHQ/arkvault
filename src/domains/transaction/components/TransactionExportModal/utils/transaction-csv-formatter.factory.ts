import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO } from "@ardenthq/sdk-profiles";
import { CURRENCIES } from "@ardenthq/sdk-intl";
import { buildTranslations } from "@/app/i18n/helpers";

const recipient = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	const { COMMON } = buildTranslations();

	if (transaction.isMultiPayment()) {
		return COMMON.MULTIPLE;
	}

	if (transaction.isTransfer()) {
		return transaction.recipient();
	}

	if (transaction.isVote() || transaction.isUnvote()) {
		return `${COMMON.VOTE} ${COMMON.TRANSACTION}`;
	}

	return COMMON.OTHER;
};

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.sender() !== transaction.wallet().address()) {
		let totalReceived = BigNumber.ZERO;

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived = totalReceived.plus(recipient.amount);
			}
		}

		return totalReceived.toNumber();
	}

	return transaction.amount();
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	const amount = transaction.isMultiPayment() ? multiPaymentAmount(transaction) : transaction.amount();

	if (transaction.isSent()) {
		return BigNumber.make(amount).times(-1).toNumber();
	}

	return amount;
};

const transactionFee = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.fee()).times(-1).toNumber();
	}

	return 0;
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.total()).times(-1).toNumber();
	}

	return transactionAmount(transaction);
};

const converted = (value: number, rate: number) => BigNumber.make(value).times(rate).toNumber();

const truncate = (value: number, currency: string) => {
	const decimals = CURRENCIES[currency]?.decimals ?? 8;

	return Math.trunc(value * 10 ** decimals) / 10 ** decimals;
};

export const CsvFormatter = (transaction: DTO.ExtendedConfirmedTransactionData, timeFormat: string) => {
	const amount = transactionAmount(transaction);
	const fee = transactionFee(transaction);
	const total = transactionTotal(transaction);

	const currency = transaction.wallet().currency();
	const exchangeCurrency = transaction.wallet().exchangeCurrency();

	const rate =
		transaction.total() === 0
			? 0
			: truncate(
					BigNumber.make(transaction.convertedTotal()).divide(transaction.total()).toNumber(),
					exchangeCurrency,
			  );

	return {
		amount: () => truncate(amount, currency),
		convertedAmount: () => truncate(converted(amount, rate), exchangeCurrency),
		convertedFee: () => (fee === 0 ? 0 : truncate(converted(fee, rate), exchangeCurrency)),
		convertedTotal: () => truncate(converted(total, rate), exchangeCurrency),
		datetime: () => transaction.timestamp()?.format(`DD.MM.YYYY ${timeFormat}`),
		fee: () => fee,
		rate: () => rate,
		recipient: () => recipient(transaction),
		sender: () => transaction.sender(),
		timestamp: () => transaction.timestamp()?.toUNIX(),
		total: () => truncate(total, currency),
	};
};
