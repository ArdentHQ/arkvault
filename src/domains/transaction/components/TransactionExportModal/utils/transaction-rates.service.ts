import { Contracts } from "@ardenthq/sdk-profiles";
import { DateTime } from "@ardenthq/sdk-intl";
import { BigNumber } from "@ardenthq/sdk-helpers";
import { HttpClient } from "@/app/services/HttpClient";

export const TransactionRates = ({ wallet }: { wallet: Contracts.IReadWriteWallet }) => {
	const exchangeEndpoint = "https://min-api.cryptocompare.com/data/histoday";
	const client = new HttpClient(0);
	const baseCurrency = "USD";

	const rates = {
		[baseCurrency]: new Map(),
	};

	const sync = async ({ from = DateTime.make(), to }: { from?: DateTime; to: DateTime }) => {
		const response = await client.get(exchangeEndpoint, {
			fsym: wallet.network().ticker(),
			limit: from.diffInDays(to),
			toTs: to.toUNIX(),
			tsym: baseCurrency,
		});

		for (const price of response.json().Data) {
			const ymd = DateTime.fromUnix(price.time).format("YYYY-MM-DD");
			rates[baseCurrency].set(ymd, BigNumber.make(price.close).toString());
		}
	};

	const byTimestamp = (timestamp?: DateTime) =>
		rates[baseCurrency].get(DateTime.make(timestamp).format("YYYY-MM-DD"));

	return {
		rate: (timestamp?: DateTime) => byTimestamp(timestamp),
		sync,
	};
};
