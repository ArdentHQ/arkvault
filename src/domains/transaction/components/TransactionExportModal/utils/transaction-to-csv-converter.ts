import { DTO } from "@ardenthq/sdk-profiles";
import { CsvSettings } from "@/domains/transaction/components/TransactionExportModal";
import { buildTranslations } from "@/app/i18n/helpers";
import { AmountFormatter } from "./transaction-amount-formatter.factory";

const headers = (settings: CsvSettings) => {
	const { COMMON } = buildTranslations();

	return [
		...(settings.includeTransactionId ? [COMMON.ID] : []),
		...(settings.includeDate ? [COMMON.TIMESTAMP, COMMON.DATE] : []),
		...(settings.includeSenderRecipient ? [COMMON.SENDER, COMMON.RECIPIENT] : []),
		...(settings.includeCryptoAmount ? [COMMON.AMOUNT, COMMON.FEE, COMMON.TOTAL] : []),
	].join(settings.delimiter);
};

const transactionToCsv = (transaction: DTO.ExtendedConfirmedTransactionData, settings: CsvSettings) => {
	const dateTimeFormat = "DD.MM.YYYY h:mm A";
	const timestamp = transaction.timestamp()?.toUNIX();
	const datetime = transaction.timestamp()?.format(dateTimeFormat);
	const sender = transaction.sender();
	const recipient = transaction.recipient();
	const amounts = AmountFormatter(transaction);

	const columns = [
		...(settings.includeTransactionId ? [transaction.id()] : []),
		...(settings.includeDate ? [timestamp, datetime] : []),
		...(settings.includeSenderRecipient ? [sender, recipient] : []),
		...(settings.includeCryptoAmount ? [amounts.amount(), amounts.fee(), amounts.total()] : []),
	];

	return columns.join(settings.delimiter);
};

export const convertToCsv = (transactions: DTO.ExtendedConfirmedTransactionData[], settings: CsvSettings) => {
	const rows = transactions.map((transaction) => transactionToCsv(transaction, settings));

	if (!settings.includeHeaderRow) {
		return rows.join("\n");
	}

	return [...[headers(settings)], ...rows].join("\n");
};
