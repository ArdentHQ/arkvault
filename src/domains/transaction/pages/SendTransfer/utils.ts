export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
