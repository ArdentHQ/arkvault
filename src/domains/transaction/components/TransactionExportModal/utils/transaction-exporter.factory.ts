import { Contracts, DTO } from "@ardenthq/sdk-profiles";
import { Services } from "@ardenthq/sdk";
import { BigNumber } from "@ardenthq/sdk-helpers";
import { useState } from "react";
import { convertToCsv } from "./transaction-to-csv-converter";
import { CsvSettings } from "@/domains/transaction/components/TransactionExportModal";
import { assertString } from "@/utils/assertions";

interface TransactionExporterFetchProperties {
	type: "all" | "received" | "sent";
	dateRange?: Services.RangeCriteria;
	cursor?: number;
}

const filterTransactions = (transactions: DTO.ExtendedConfirmedTransactionData[]) =>
	transactions.filter((transaction) => {
		if (transaction.isTransfer()) {
			return transaction.sender() !== transaction.recipient();
		}

		if (transaction.isMultiPayment()) {
			let amount = BigNumber.make(transaction.amount());

			for (const recipient of transaction.recipients()) {
				if (transaction.sender() === recipient.address) {
					amount = amount.minus(recipient.amount);
				}
			}

			return !amount.isZero();
		}

		return false;
	});

export const TransactionExporter = ({
	profile,
	wallet,
	limit = 100,
}: {
	profile: Contracts.IProfile;
	wallet: Contracts.IReadWriteWallet;
	limit?: number;
}) => {
	const exchangeCurrency = profile.settings().get<string>(Contracts.ProfileSetting.ExchangeCurrency);
	const timeFormat = profile.settings().get<string>(Contracts.ProfileSetting.TimeFormat);

	let count = 0;

	assertString(exchangeCurrency);
	assertString(timeFormat);

	let requestedSyncAbort = false;

	let transactions: DTO.ExtendedConfirmedTransactionData[] = [];

	const sync = async ({
		type = "all",
		dateRange,
		cursor = 1,
	}: TransactionExporterFetchProperties): Promise<number | undefined> => {
		// Clear cache.
		if (cursor === 1) {
			transactions = [];
			requestedSyncAbort = false;
		}

		if (requestedSyncAbort) {
			return;
		}

		const page = await wallet.transactionIndex()[type]({ cursor, limit, timestamp: dateRange });

		transactions.push(...page.items());
		cursor = cursor + 1;

		count = transactions.length;

		// Last page.
		// TODO: Not relying on totalCount because it is an estimate
		//        and is not giving accurate pagination info. Address this issue after initial implementation.
		if (page.items().length < limit) {
			if (type === "received") {
				transactions = filterTransactions(transactions);
			}

			return transactions.length;
		}

		return sync({ cursor, dateRange, type });
	};

	return {
		transactions: () => ({
			abortSync: () => (requestedSyncAbort = true),
			count: () => count,
			items: () => transactions,
			sync,
			toCsv: (settings: CsvSettings) => convertToCsv(transactions, settings, exchangeCurrency, timeFormat),
		}),
	};
};
