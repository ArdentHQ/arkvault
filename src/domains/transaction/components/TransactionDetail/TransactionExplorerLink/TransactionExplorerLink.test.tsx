import React from "react";

import * as reactResizeDetector from "react-resize-detector";
import { TransactionExplorerLink } from "./TransactionExplorerLink";
import { translations } from "@/domains/transaction/i18n";
import { render } from "@/utils/testing-library";

describe("TransactionExplorerLink", () => {
	it("should render a transaction link", () => {
		const { container } = render(
			<TransactionExplorerLink
				// @ts-ignore
				transaction={{
					explorerLink: () => "transaction-link",
					id: () => "test-id",
				}}
			/>,
		);

		expect(container).toHaveTextContent(translations.ID);
		expect(container).toMatchSnapshot();
	});

	it("should render a transaction link responsive after the width has calculated", () => {
		vi.spyOn(reactResizeDetector, "useResizeDetector").mockReturnValue({ width: 100 });

		const { container } = render(
			<TransactionExplorerLink
				// @ts-ignore
				transaction={{
					explorerLink: () => "transaction-link",
					id: () => "test-id",
				}}
			/>,
		);

		expect(container).toHaveTextContent(translations.ID);
		expect(container).toMatchSnapshot();
	});
});
