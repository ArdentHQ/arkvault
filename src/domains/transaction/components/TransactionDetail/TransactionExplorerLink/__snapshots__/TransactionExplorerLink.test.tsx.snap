// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionExplorerLink > should render a transaction link 1`] = `
<div>
  <div
    class="no-ligatures css-qu17s6"
    data-testid="TransactionDetail"
  >
    <div
      class="w-40 flex-1 space-y-2 whitespace-nowrap"
      data-testid="TransactionDetail--desktop"
    >
      <div
        class="no-ligatures text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
      >
        ID
      </div>
      <div
        class="flex items-center font-semibold"
      >
        <div
          class="flex w-full items-center space-x-3 overflow-hidden"
        >
          <span
            class="overflow-hidden"
          >
            <a
              class="ring-focus group/inner inline-block flex items-center css-7nwwrh"
              data-ring-focus-margin="-m-1"
              data-testid="Link"
              href="transaction-link"
              rel="noopener noreferrer"
            >
              <span
                class="css-bxx244"
              >
                <span
                  class="no-ligatures css-0"
                  data-testid="TruncateMiddle"
                >
                  test-id
                </span>
              </span>
              <div
                class="mb-[3px] shrink-0 align-middle duration-200 inline-block text-sm css-v0ob3f"
                data-testid="Link__external"
                height="16"
                width="16"
              >
                <svg
                  id="arrow-external"
                  viewBox="0 0 20 20"
                  x="0"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0"
                >
                  <path
                    d="M18.999 6.43V1h-5.43m5.43 0L7.033 12.966M9.749 4h-7.5c-.625 0-1.25.625-1.25 1.25v12.5c0 .625.625 1.25 1.25 1.25h12.5c.625 0 1.25-.625 1.25-1.25v-7.5"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </a>
          </span>
          <span
            class="flex text-theme-primary-300 dark:text-theme-secondary-600"
          >
            <button
              class="ring-focus relative focus:outline-none"
              data-ring-focus-margin="-m-1"
              data-testid="clipboard-icon__wrapper"
              type="button"
            >
              <div
                class="css-v0ob3f"
                height="16"
                width="16"
              >
                <svg
                  id="copy"
                  viewBox="0 0 20 20"
                  x="0"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0"
                >
                  <g
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  >
                    <path
                      d="M7 17.765C7 18.485 7.496 19 8.09 19h-.099 8.82c.694 0 1.189-.618 1.189-1.235V6.235C18 5.515 17.405 5 16.81 5h0-6.242c-.298 0-.595.103-.892.309l-2.379 2.47C7 8.09 7 8.397 7 8.706v9.059z"
                    />
                    <path
                      d="M13 5V2.235C13 1.515 12.405 1 11.81 1h0-6.242c-.298 0-.595.103-.892.309l-2.379 2.47C2 4.09 2 4.397 2 4.706v9.059C2 14.485 2.496 15 3.09 15h-.099H7"
                    />
                  </g>
                </svg>
              </div>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TransactionExplorerLink > should render a transaction link responsive after the width has calculated 1`] = `
<div>
  <div
    class="no-ligatures css-qu17s6"
    data-testid="TransactionDetail"
  >
    <div
      class="w-40 flex-1 space-y-2 whitespace-nowrap"
      data-testid="TransactionDetail--desktop"
    >
      <div
        class="no-ligatures text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
      >
        ID
      </div>
      <div
        class="flex items-center font-semibold"
      >
        <div
          class="flex w-full items-center space-x-3 overflow-hidden"
        >
          <span
            class="overflow-hidden"
          >
            <a
              class="ring-focus group/inner inline-block flex items-center css-7nwwrh"
              data-ring-focus-margin="-m-1"
              data-testid="Link"
              href="transaction-link"
              rel="noopener noreferrer"
            >
              <span
                class="css-bxx244"
              >
                <span
                  class="no-ligatures css-0"
                  data-testid="TruncateMiddle"
                >
                  test-id
                </span>
              </span>
              <div
                class="mb-[3px] shrink-0 align-middle duration-200 inline-block text-sm css-v0ob3f"
                data-testid="Link__external"
                height="16"
                width="16"
              >
                <svg
                  id="arrow-external"
                  viewBox="0 0 20 20"
                  x="0"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0"
                >
                  <path
                    d="M18.999 6.43V1h-5.43m5.43 0L7.033 12.966M9.749 4h-7.5c-.625 0-1.25.625-1.25 1.25v12.5c0 .625.625 1.25 1.25 1.25h12.5c.625 0 1.25-.625 1.25-1.25v-7.5"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </a>
          </span>
          <span
            class="flex text-theme-primary-300 dark:text-theme-secondary-600"
          >
            <button
              class="ring-focus relative focus:outline-none"
              data-ring-focus-margin="-m-1"
              data-testid="clipboard-icon__wrapper"
              type="button"
            >
              <div
                class="css-v0ob3f"
                height="16"
                width="16"
              >
                <svg
                  id="copy"
                  viewBox="0 0 20 20"
                  x="0"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0"
                >
                  <g
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  >
                    <path
                      d="M7 17.765C7 18.485 7.496 19 8.09 19h-.099 8.82c.694 0 1.189-.618 1.189-1.235V6.235C18 5.515 17.405 5 16.81 5h0-6.242c-.298 0-.595.103-.892.309l-2.379 2.47C7 8.09 7 8.397 7 8.706v9.059z"
                    />
                    <path
                      d="M13 5V2.235C13 1.515 12.405 1 11.81 1h0-6.242c-.298 0-.595.103-.892.309l-2.379 2.47C2 4.09 2 4.397 2 4.706v9.059C2 14.485 2.496 15 3.09 15h-.099H7"
                    />
                  </g>
                </svg>
              </div>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
